src/garage/envs/garage_env.py

```diff
@@ -92,34 +106,154 @@ def __init__(self, env=None, env_name='', is_image=False):
+                             max_episode_length=max_episode_length)
+
+    @property
+    def action_space(self):
+        """akro.Space: The action space specification."""
+        return self._action_space
+
+    @property
+    def observation_space(self):
+        """akro.Space: The observation space specification."""
+        return self._observation_space
 
     @property
     def spec(self):
-        """Return the environment specification.
+        """garage.envs.env_spec.EnvSpec: The envionrment specification."""
+        return self._spec
+
+    @property
+    def render_modes(self):
+        """list: A list of string representing the supported render modes."""
+        return self._render_modes
 
-        This property needs to exist, since it's defined as a property in
-        gym.Wrapper in a way that makes it difficult to overwrite.
+    def reset(self, **kwargs):
+        """Call reset on wrapped env.
+
+        Args:
+            kwargs: Keyword args
 
         Returns:
-            garage.envs.env_spec.EnvSpec: The envionrment specification.
+            numpy.ndarray: The first observation. It must conforms to
+            `observation_space`.
+            dict: The episode-level information. Note that this is not part
+            of `env_info` provided in `step()`. It contains information of
+            the entire episode， which could be needed to determine the first
+            action (e.g. in the case of goal-conditioned or MTRL.)
 
         """
-        return self._spec
+        first_obs = self.env.reset(**kwargs)
+
+        self._step_cnt = 0
+        self._last_observation = first_obs
+        # Populate episode_info if needed.
+        episode_info = {}
+        return first_obs, episode_info
+
+    def step(self, action):
+        """Call step on wrapped env.
+
+        Args:
+            action (np.ndarray): An action provided by the agent.
+
+        Returns:
+            TimeStep: The time step resulting from the action.
+
+        Raises:
+            RuntimeError: if `step()` is called after the environment has been
+                constructed and `reset()` has not been called.
+
+        """
+        if self._last_observation is None:
+            raise RuntimeError('reset() must be called before step()!')
+
+        observation, reward, done, info = self.env.step(action)
+
+        if self._visualize:
+            self.env.render(mode='human')
+
+        last_obs = self._last_observation
+        # Type conversion
+        if not isinstance(reward, float):
+            reward = float(reward)
+
+        self._last_observation = observation
+        self._step_cnt += 1
+
+        step_type = None
+        if done:
+            step_type = StepType.TERMINAL
+        elif self._step_cnt == 1:
+            step_type = StepType.FIRST
+        else:
+            step_type = StepType.MID
+
+        # gym envs that are wrapped in TimeLimit wrapper modify
```

Comment:
is  there a reason a lot of this logic is duplicated between `BulletEnv` and here? can you subclass `BulletEnv` from `GymEnv`?

Reply:
Can I? My concern is `GymEnv` creates a `BulletEnv` in `__new__()`. Can a parent class returns a subclass? Searched for this and couldn't find good explanation. (Or maybe I am not searching the right key words?)

If this is possible, then yes, `BulletEnv` should be basically the same except `__init__` and `get/setstate()`.

Reply:
In Python, you can do whatever you want :).